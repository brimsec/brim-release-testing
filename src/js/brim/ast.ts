import lib from "../lib"

export default function ast(tree: any) {
  return {
    valid() {
      return !tree.error
    },
    error() {
      return tree.error || null
    },
    groupByKeys() {
      const g = this.proc("GroupByProc")
      const keys = g ? g.keys : []
      return keys.map((k) => k.target)
    },
    proc(name: string) {
      return getProcs(tree).find((p) => p.op === name)
    },
    procs(name: string): any[] {
      return getProcs(tree).filter((p) => p.op === name)
    },
    getProcs() {
      return getProcs(tree)
    },
    self() {
      return tree
    },
    sorts() {
      return this.procs("SortProc").reduce((sorts, proc) => {
        lib.array.wrap(proc.fields).forEach((field) => {
          sorts[fieldExprToName(field)] = proc.sortdir === 1 ? "asc" : "desc"
        })
        return sorts
      }, {})
    }
  }
}

function fieldExprToName(expr) {
  switch (expr.op) {
    case "FieldRead":
      return expr.field
    case "FieldCall":
      return `${fieldExprToName(expr.field)}.${expr.param}`
    case "Index":
      return `${fieldExprToName(expr.field)}[${expr.param}]`
    default:
      // XXX
      return ""
  }
}

function getProcs(ast) {
  if (!ast || ast.error) return []
  const list = []
  collectProcs(ast, list)
  return list
}

function collectProcs(proc, list) {
  if (COMPOUND_PROCS.includes(proc.op)) {
    for (const p of proc.procs) collectProcs(p, list)
  } else {
    list.push(proc)
  }
}

export const HEAD_PROC = "HeadProc"
export const TAIL_PROC = "TailProc"
export const SORT_PROC = "SortProc"
export const FILTER_PROC = "FilterProc"
export const PARALLEL_PROC = "ParallelProc"
export const SEQUENTIAL_PROC = "SequentialProc"
export const SOURCE_PROC = "SourceProc"
export const TUPLE_PROCS = [
  SOURCE_PROC,
  HEAD_PROC,
  TAIL_PROC,
  SORT_PROC,
  FILTER_PROC,
  SEQUENTIAL_PROC
]
export const COMPOUND_PROCS = [PARALLEL_PROC, SEQUENTIAL_PROC]
export const EVERYTHING_FILTER = {op: "MatchAll"}
