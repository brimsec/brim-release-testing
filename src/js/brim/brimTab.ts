import {get} from "lodash"

import {SpacesState} from "../state/Spaces/types"
import {TabState} from "../state/Tab/types"
import lib from "../lib"

export default function(tab: TabState, spaces: SpacesState) {
  return {
    title() {
      const name = get(
        spaces,
        [tab.current.connectionId || "", tab.current.spaceId || "", "name"],
        "New Tab"
      )
      return lib.compact([name, tab.searchBar.previous]).join(": ")
    }
  }
}
