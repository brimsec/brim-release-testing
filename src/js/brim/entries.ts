import isEqual from "lodash/isEqual"

type Args = {
  position: number
  entries: any[]
}

export default function({entries: initEntries, position}: Args) {
  const entries = [...initEntries]

  if (position < -1 || position >= entries.length) {
    throw new Error("Position out of bounds")
  }

  return {
    push(entry: any) {
      if (!isEqual(entry, this.getCurrentEntry())) {
        entries.splice(position + 1, entries.length, entry)
        position = entries.length - 1
      }
      return this
    },
    update(updates: any) {
      entries[position] = {...entries[position], ...updates}
      return this
    },
    goBack() {
      if (this.canGoBack()) position -= 1
      return this
    },
    goForward() {
      if (this.canGoForward()) position += 1
      return this
    },
    canGoBack() {
      return entries.length > 1 ? position !== 0 : false
    },
    canGoForward() {
      return entries.length > 1 ? position < entries.length - 1 : false
    },
    getEntries() {
      return entries
    },
    getCurrentEntry() {
      return entries[position]
    },
    data() {
      return {position, entries}
    }
  }
}
