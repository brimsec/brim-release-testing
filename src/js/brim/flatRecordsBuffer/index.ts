import {isArray} from "../../lib/is"
import channel, {$Channel} from "./channel"
import nestedRecord from "./nestedRecord"
import zngToZeekTypes from "./zngToZeekTypes"

export type RawColumn = {name: string; type: string | RawColumn[]}
export type RawValue = string | null | RawValue[]
export type RawRecord = {
  id: number
  type?: RawColumn[]
  values: RawValue[]
}

export default function flatRecordBuffers() {
  const channels = new Map<number, $Channel>()
  const types = {}

  return {
    getChannel(id: number) {
      let chan = channels.get(id)
      if (chan) {
        return chan
      } else {
        chan = channel(id)
        channels.set(id, chan)
        return chan
      }
    },

    add(id: number, records: RawRecord[]) {
      const chan = this.getChannel(id)
      const flatRecords = records.map((r) => {
        if (r.type) types[r.id] = zngToZeekTypes(r.type)
        return nestedRecord(r.values, types[r.id]).flatten()
      })
      chan.add(flatRecords)
    },

    empty(chanId = 0) {
      return this.getChannel(chanId).empty()
    },

    columns() {
      return Object.keys(types).reduce((all, id) => {
        all[id] = flattenType(types[id])
        return all
      }, {})
    },

    channels() {
      return Array.from<$Channel>(channels.values())
    },

    clearRecords(chanId = 0) {
      this.getChannel(chanId).clear()
    }
  }
}

function flattenType(descriptor, prefix = "") {
  return descriptor.reduce((flat, {name, type}) => {
    const cols = isArray(type)
      ? flattenType(type, `${prefix}${name}.`)
      : [{name: prefix + name, type}]

    return flat.concat(cols)
  }, [])
}
