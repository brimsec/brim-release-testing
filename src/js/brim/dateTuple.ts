import {DateTuple} from "../lib/TimeWindow"
import brim, {Span} from "./"

export default function dateTuple(dt: DateTuple) {
  return {
    toSpan(): Span {
      return [brim.time(dt[0]).toTs(), brim.time(dt[1]).toTs()]
    }
  }
}
