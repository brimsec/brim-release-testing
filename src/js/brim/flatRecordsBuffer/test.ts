import brim from ".."

test("coverts to an array of records", () => {
  const buffer = brim.flatRecordsBuffer()
  buffer.add(0, records)
  expect(buffer.channels()[0].records()).toMatchSnapshot()
})

test("flattens the columns", () => {
  const buffer = brim.flatRecordsBuffer()
  buffer.add(0, nestedRecords)
  expect(buffer.columns()).toMatchSnapshot()
})

test("coverts to an array of nested records", () => {
  const buffer = brim.flatRecordsBuffer()
  buffer.add(0, nestedRecords)
  expect(buffer.channels()[0].records()).toMatchSnapshot()
})

test("triple nested", () => {
  const buffer = brim.flatRecordsBuffer()
  buffer.add(0, tripleNest)
  expect(buffer.channels()[0].records()).toEqual([
    [
      {
        name: "a.b.c",
        type: "addr",
        value: "192.168.0.1"
      }
    ]
  ])
})

const records = [
  {
    id: 0,
    type: [
      {name: "ts", type: "time"},
      {name: "_path", type: "string"},
      {name: "count", type: "count"}
    ],
    values: ["1428917670.000000000", "dns", "2"]
  },
  {id: 0, values: ["1428917670.000000000", "conn", "3"]},
  {id: 0, values: ["1428917670.000000000", "capture_loss", "1"]},
  {id: 0, values: ["1428917640.000000000", "conn", "1"]}
]

const nestedRecords = [
  {
    id: 10,
    type: [
      {name: "_path", type: "string"},
      {name: "ts", type: "time"},
      {name: "uid", type: "string"},
      {
        name: "id",
        type: [
          {name: "orig_h", type: "addr"},
          {name: "orig_p", type: "port"},
          {name: "resp_h", type: "addr"},
          {name: "resp_p", type: "port"}
        ]
      },
      {name: "proto", type: "enum"},
      {name: "service", type: "string"},
      {name: "duration", type: "interval"},
      {name: "orig_bytes", type: "count"},
      {name: "resp_bytes", type: "count"},
      {name: "conn_state", type: "string"},
      {name: "local_orig", type: "bool"},
      {name: "local_resp", type: "bool"},
      {name: "missed_bytes", type: "count"},
      {name: "history", type: "string"},
      {name: "orig_pkts", type: "count"},
      {name: "orig_ip_bytes", type: "count"},
      {name: "resp_pkts", type: "count"},
      {name: "resp_ip_bytes", type: "count"},
      {name: "tunnel_parents", type: "set[string]"}
    ],
    values: [
      "conn",
      "1428917684.732640",
      "ClM1vE3c4iNZUgpZhh",
      ["fe80::eef4:bbff:fe4f:b245", "143", "ff02::16", "0"],
      "icmp",
      null,
      null,
      null,
      null,
      "OTH",
      "F",
      "F",
      "0",
      null,
      "1",
      "76",
      "0",
      "0",
      null
    ]
  }
]

const tripleNest = [
  {
    id: 0,
    type: [{name: "a", type: [{name: "b", type: [{name: "c", type: "addr"}]}]}],
    values: [[["192.168.0.1"]]]
  }
]
