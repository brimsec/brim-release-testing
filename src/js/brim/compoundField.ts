import {FieldValue} from "../types/records"
import {ONE_CHAR} from "./field"
import {isArray} from "../lib/is"
import brim from "./"

export const COMPOUND_FIELD_RGX = /^(set|array)\[(\w+)\]$/

function compoundField(
  name: string,
  type: string,
  value: FieldValue[] | null | undefined
) {
  const match = type.match(COMPOUND_FIELD_RGX)
  if (!match) throw new Error("Not compound type: " + type)
  const [_, container, itemType] = match
  let items = []
  if (isArray(value)) items = value

  return {
    name,
    container,
    itemType,
    length: items.length,
    items() {
      return items.map((_, index) => this.item(index))
    },
    item(index: number) {
      if (items[index]) {
        return brim.field({name, type: itemType, value: items[index]})
      } else {
        return null
      }
    },
    guessWidth() {
      const comma = ONE_CHAR
      const wrap = 2 * ONE_CHAR
      const items = this.items()
      let sum = 0
      for (const item of items) {
        sum += item.guessWidth()
      }
      sum += comma * (items.length - 1)
      sum += wrap
      return sum
    }
  }
}

export default compoundField
