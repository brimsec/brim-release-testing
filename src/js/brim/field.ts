import {isEqual} from "lodash"

import {COMPOUND_FIELD_RGX} from "./compoundField"
import {FieldData} from "../types/records"
import {withCommas} from "../lib/fmt"
import brim, {$Field} from "./"

export const ONE_CHAR = 7.39
export const FIELD_PAD = 14
export const PATH_PAD = 12
const WHITE_SPACE = /\s+/
const COMMA = /,/
const STRING_TYPE = /^b?string$/
const DOUBLE_QUOTE = /"/g
const ESCAPED_DOUBLE_QUOTE = '\\"'
const BACK_SLASH = /\\/g
const ESCAPED_BACK_SLASH = "\\\\"

function field({name, type, value}: FieldData): $Field {
  return {
    name,
    type,
    value,
    queryableValue() {
      if (this.compound()) {
        return this.toCompound()
          .items()
          .map(field)
          .map((f) => f.queryableValue())
          .join(" ")
      }
      if (this.value === null) return "null"
      if (this.type === "bool") return this.value === "T" ? "true" : "false"
      let quote = [WHITE_SPACE, COMMA].some((reg) => reg.test(this.value))
      if (STRING_TYPE.test(this.type)) quote = true
      const str = this.value
        .replace(BACK_SLASH, ESCAPED_BACK_SLASH)
        .replace(DOUBLE_QUOTE, ESCAPED_DOUBLE_QUOTE)
      return quote ? `"${str}"` : str
    },
    stringValue(): string {
      if (value === null) return "null"
      else if (Array.isArray(value)) return value.join(",")
      else return value
    },

    compound() {
      return COMPOUND_FIELD_RGX.test(type)
    },
    toCompound() {
      // @ts-ignore
      return brim.compoundField(name, type, value)
    },
    toDate() {
      return new Date(+this.value * 1000)
    },
    display() {
      if (value === "(empty)") {
        return ""
      } else if (value === null) {
        return "â¦»"
      } else if (type === "count") {
        return withCommas(value.toString())
      } else if (type === "time") {
        return brim.time(this.toDate()).format()
      } else if (isEqual(value, {})) {
        return ""
      } else {
        return value.toString()
      }
    },
    guessWidth() {
      if (name === "_path") {
        return this.display().length * ONE_CHAR + FIELD_PAD + PATH_PAD
      } else if (this.compound()) {
        return this.toCompound().guessWidth()
      } else {
        return Math.ceil(this.display().length * ONE_CHAR + FIELD_PAD)
      }
    }
  }
}

export default field
