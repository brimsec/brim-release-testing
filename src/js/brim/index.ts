import {FieldValue} from "../types/records"
import ast from "./ast"
import compoundField from "./compoundField"
import dateTuple from "./dateTuple"
import entries from "./entries"
import field from "./field"
import flatRecordsBuffer from "./flatRecordsBuffer"
import form from "./form"
import interop from "./interop"
import log from "./log"
import program from "./program"
import randomHash from "./randomHash"
import record from "./record"
import relTime from "./relTime"
import space from "./space"
import span from "./span"
import syntax from "./syntax"
import tab from "./brimTab"
import table from "./table"
import time from "./time"
import zeekLogInfo from "./zeekLogInfo"

export type $Field = {
  name: string
  type: string
  value: FieldValue
  stringValue: () => string
  queryableValue: () => string
  compound: () => boolean
  toCompound: () => $CompoundField
  toDate: () => Date
  display: () => string
  guessWidth: () => number
}

export type $CompoundField = ReturnType<typeof compoundField>
export type $Log = ReturnType<typeof log>

export type Ts = {
  ns: number
  sec: number
}

export type BrimSpace = ReturnType<typeof space>

export type Span = [Ts, Ts]

export default {
  table,
  program,
  field,
  compoundField,
  log,
  ast,
  syntax,
  space,
  time,
  relTime,
  span,
  dateTuple,
  flatRecordsBuffer,
  form,
  record,
  interop,
  randomHash,
  entries,
  tab,
  zeekLogInfo
}
