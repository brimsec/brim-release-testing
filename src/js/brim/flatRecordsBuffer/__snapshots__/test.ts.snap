// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coverts to an array of nested records 1`] = `
Array [
  Array [
    Object {
      "name": "_path",
      "type": "string",
      "value": "conn",
    },
    Object {
      "name": "ts",
      "type": "time",
      "value": "1428917684.732640",
    },
    Object {
      "name": "uid",
      "type": "string",
      "value": "ClM1vE3c4iNZUgpZhh",
    },
    Object {
      "name": "id.orig_h",
      "type": "addr",
      "value": "fe80::eef4:bbff:fe4f:b245",
    },
    Object {
      "name": "id.orig_p",
      "type": "port",
      "value": "143",
    },
    Object {
      "name": "id.resp_h",
      "type": "addr",
      "value": "ff02::16",
    },
    Object {
      "name": "id.resp_p",
      "type": "port",
      "value": "0",
    },
    Object {
      "name": "proto",
      "type": "enum",
      "value": "icmp",
    },
    Object {
      "name": "service",
      "type": "string",
      "value": null,
    },
    Object {
      "name": "duration",
      "type": "interval",
      "value": null,
    },
    Object {
      "name": "orig_bytes",
      "type": "count",
      "value": null,
    },
    Object {
      "name": "resp_bytes",
      "type": "count",
      "value": null,
    },
    Object {
      "name": "conn_state",
      "type": "string",
      "value": "OTH",
    },
    Object {
      "name": "local_orig",
      "type": "bool",
      "value": "F",
    },
    Object {
      "name": "local_resp",
      "type": "bool",
      "value": "F",
    },
    Object {
      "name": "missed_bytes",
      "type": "count",
      "value": "0",
    },
    Object {
      "name": "history",
      "type": "string",
      "value": null,
    },
    Object {
      "name": "orig_pkts",
      "type": "count",
      "value": "1",
    },
    Object {
      "name": "orig_ip_bytes",
      "type": "count",
      "value": "76",
    },
    Object {
      "name": "resp_pkts",
      "type": "count",
      "value": "0",
    },
    Object {
      "name": "resp_ip_bytes",
      "type": "count",
      "value": "0",
    },
    Object {
      "name": "tunnel_parents",
      "type": "set[string]",
      "value": null,
    },
  ],
]
`;

exports[`coverts to an array of records 1`] = `
Array [
  Array [
    Object {
      "name": "ts",
      "type": "time",
      "value": "1428917670.000000000",
    },
    Object {
      "name": "_path",
      "type": "string",
      "value": "dns",
    },
    Object {
      "name": "count",
      "type": "count",
      "value": "2",
    },
  ],
  Array [
    Object {
      "name": "ts",
      "type": "time",
      "value": "1428917670.000000000",
    },
    Object {
      "name": "_path",
      "type": "string",
      "value": "conn",
    },
    Object {
      "name": "count",
      "type": "count",
      "value": "3",
    },
  ],
  Array [
    Object {
      "name": "ts",
      "type": "time",
      "value": "1428917670.000000000",
    },
    Object {
      "name": "_path",
      "type": "string",
      "value": "capture_loss",
    },
    Object {
      "name": "count",
      "type": "count",
      "value": "1",
    },
  ],
  Array [
    Object {
      "name": "ts",
      "type": "time",
      "value": "1428917640.000000000",
    },
    Object {
      "name": "_path",
      "type": "string",
      "value": "conn",
    },
    Object {
      "name": "count",
      "type": "count",
      "value": "1",
    },
  ],
]
`;

exports[`flattens the columns 1`] = `
Object {
  "10": Array [
    Object {
      "name": "_path",
      "type": "string",
    },
    Object {
      "name": "ts",
      "type": "time",
    },
    Object {
      "name": "uid",
      "type": "string",
    },
    Object {
      "name": "id.orig_h",
      "type": "addr",
    },
    Object {
      "name": "id.orig_p",
      "type": "port",
    },
    Object {
      "name": "id.resp_h",
      "type": "addr",
    },
    Object {
      "name": "id.resp_p",
      "type": "port",
    },
    Object {
      "name": "proto",
      "type": "enum",
    },
    Object {
      "name": "service",
      "type": "string",
    },
    Object {
      "name": "duration",
      "type": "interval",
    },
    Object {
      "name": "orig_bytes",
      "type": "count",
    },
    Object {
      "name": "resp_bytes",
      "type": "count",
    },
    Object {
      "name": "conn_state",
      "type": "string",
    },
    Object {
      "name": "local_orig",
      "type": "bool",
    },
    Object {
      "name": "local_resp",
      "type": "bool",
    },
    Object {
      "name": "missed_bytes",
      "type": "count",
    },
    Object {
      "name": "history",
      "type": "string",
    },
    Object {
      "name": "orig_pkts",
      "type": "count",
    },
    Object {
      "name": "orig_ip_bytes",
      "type": "count",
    },
    Object {
      "name": "resp_pkts",
      "type": "count",
    },
    Object {
      "name": "resp_ip_bytes",
      "type": "count",
    },
    Object {
      "name": "tunnel_parents",
      "type": "set[string]",
    },
  ],
}
`;
